node/src/tasks/bsp_charge_fees.rs

```diff
@@ -104,3 +198,197 @@ where
+            let stop_storing_for_insolvent_user_call = storage_hub_runtime::RuntimeCall::FileSystem(
+                pallet_file_system::Call::stop_storing_for_insolvent_user {
+                    file_key: *file_key,
+                    bucket_id,
+                    location,
+                    owner,
+                    fingerprint,
+                    size,
+                    inclusion_forest_proof,
+                },
+            );
+
+            // Send the confirmation transaction and wait for it to be included in the block and
+            // continue only if it is successful.
+            self.storage_hub_handler
+                .blockchain
+                .send_extrinsic(stop_storing_for_insolvent_user_call, Tip::from(0))
+                .await?
+                .with_timeout(Duration::from_secs(
+                    self.storage_hub_handler
+                        .provider_config
+                        .extrinsic_retry_timeout,
+                ))
+                .watch_for_success(&self.storage_hub_handler.blockchain)
+                .await?;
+
+            trace!(target: LOG_TARGET, "Stop storing submitted successfully");
+
+            // Remove the file from the forest.
+            self.remove_file_from_forest(&file_key).await?;
+
+            // Check that the new Forest root matches the one on-chain.
+            let own_provider_id = match self
+                .storage_hub_handler
+                .blockchain
+                .query_storage_provider_id(None)
+                .await?
+                .ok_or(anyhow!("Failed to get own provider ID"))?
+            {
+                StorageProviderId::MainStorageProvider(id)
+                | StorageProviderId::BackupStorageProvider(id) => id,
+            };
+
+            self.check_provider_root(own_provider_id.into()).await?;
+
+            // If that was the last file of the user then charge the user for the debt they have.
+            if user_files.len() == 1 {
+                let call = storage_hub_runtime::RuntimeCall::PaymentStreams(
+                    pallet_payment_streams::Call::charge_payment_streams {
+                        user_account: insolvent_user,
+                    },
+                );
+
+                let charging_result = self
+                    .storage_hub_handler
+                    .blockchain
+                    .send_extrinsic(call, Tip::from(0))
+                    .await;
+
+                match charging_result {
+                    Ok(submitted_transaction) => {
+                        info!(target: LOG_TARGET, "Submitted extrinsic to charge users with debt: {}", submitted_transaction.hash());
+                    }
+                    Err(e) => {
+                        error!(target: LOG_TARGET, "Failed to send extrinsic to charge users with debt: {}", e);
+                    }
+                }
+            }
+        }
+
+        // Release the forest root write "lock".
```

Comment:
I'd at least log if there is an error here. It would be critical.

Reply:
Yeah I saw this done in a few other tasks so I did the same but I was sceptical of not managing the result of the actual forest write. I'll update it in all tasks